lib/rmt/downloader.rb

```diff
@@ -89,72 +83,104 @@ def create_fiber_request(local_filenames, file_reference, failed_downloads: nil)
-
     request = RMT::FiberRequest.new(
-      uri.to_s,
+      request_uri(file).to_s,
       download_path: downloaded_file,
       request_fiber: request_fiber,
-      remote_file: file.remote_path,
-      followlocation: true,
-      headers: headers
+      followlocation: true
     )
 
     request.receive_headers
     request.receive_body
   end
 
+  def try_copying_from_cache(files, ignore_errors: false)
+    uncached_files = []
+    cacheable_files = {}
+    files.each do |file|
+      # RMT must not make HEAD requests when importing repos (file://)
+      next uncached_files << file if file.remote_path.scheme == 'file'
+      next uncached_files << file if file.cache_timestamp.nil?
+
+      cacheable_files[file] = head_request(file)
+    end
+
+    return [uncached_files, []] if cacheable_files.empty?
+
+    if cacheable_files.count > 1
+      hydra = Typhoeus::Hydra.new(max_concurrency: @concurrency)
+      cacheable_files.each_value { |request| hydra.queue(request) }
+      hydra.run
+    else
+      cacheable_files.values.first.run
+    end
+
+    failed_files = []
+    cacheable_files.each do |(file, request)|
+      next uncached_files << file unless valid_cached_file?(file, request.response)
+
+      copy_from_cache(file)
+    rescue RMT::Downloader::Exception => e
+      next failed_files << file.local_path if ignore_errors
+
+      raise e
+    end
+
+    [uncached_files, failed_files]
+  end
+
+  def head_request(file)
+    RMT::HttpRequest.new(request_uri(file).to_s, method: :head, followlocation: true)
+  end
+
+  def valid_cached_file?(file, response)
+    raise http_error(file.remote_path, response.code) if (response.code != 200)
+
+    # response.headers returns Typhoeus::Response::Headers, which takes care of
+    # case-sensitive concerns with the header's key
+    last_modified_header = response.headers['Last-Modified']
+    return false unless last_modified_header
+
+    file.cache_timestamp == Time.parse(last_modified_header).utc
+  end
+
   def copy_from_cache(file)
+    make_file_dir(file.local_path)
     FileUtils.cp(file.cache_path, file.local_path, preserve: true) unless (file.cache_path == file.local_path)
     @logger.info("→ #{File.basename(file.local_path)}")
   end
 
   def finalize_download(request, file)
     if (URI(request.base_url).scheme != 'file' && request.response.code != 200)
-      raise RMT::Downloader::Exception.new(
-        _('%{file} - HTTP request failed with code %{code}') % {
-          file: request.remote_file,
-          code: request.response.code
-        },
-        request.response.code
-      )
+      raise http_error(request.remote_file, request.response.code)
     end
 
     handle_checksum_verification!(file.checksum_type, file.checksum, request.download_path)
 
     FileUtils.mv(request.download_path.path, file.local_path)
     File.chmod(0o644, file.local_path)
+    if (last_modified = request.response.headers['Last-Modified'])
```

Comment:
I think that this if statement could become confusing because the `if something = something-else` looks like a bug and hides the intention of the statement.